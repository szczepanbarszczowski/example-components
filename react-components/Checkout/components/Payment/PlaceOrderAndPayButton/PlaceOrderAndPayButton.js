import React, { useMemo, useCallback } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useTranslation } from 'react-i18next'
import {
  lt,
  identity,
  any,
  keys,
  pipe,
  length,
  equals,
  not,
  always
} from 'ramda'
import { formValueSelector } from 'redux-form'
import { createSelector } from 'reselect'
import Button from 'components/Buttons/Button'

import { checkoutCardValidator, checkoutCODValidator } from 'app/validators'
import {
  selectCheckoutState,
  selectSelectedPaymentMethod,
  selectIsCheckoutPageLocked,
  selectIsCheckoutPageLoading,
  isPayByPayFort,
  selectSelectedPayment
} from 'selectors/stateSelectors/checkoutSelectors'

import {
  selectCalculationShipmentMethods,
  selectCalculationProducts
} from 'selectors/apiSelectors/calculationSelectors'
import { CheckoutApiActions } from 'actions/apiActions/checkoutApiActions'
import { CheckoutActions } from 'actions/stateActions/checkoutActions'
import { selectUserCreditCardProviderById } from 'selectors/userSelectors'

import {
  mapCheckoutDataForCardPayment,
  mapCheckoutDataForCOD
} from 'app/redux/services'

import { paymentMethods, payFortProvider } from 'reduxConstants/checkout'
import withStyles from 'lib/withStyles'
import { isPayfort } from 'services/configurationCheck'
import { discountCode as form } from 'constants/formNames'
import s from './PlaceOrderAndPayButton.module.scss'

export const hasCheckoutErrorsSelector = createSelector(
  selectCalculationProducts,
  selectCalculationShipmentMethods,
  selectCheckoutState,
  selectSelectedPaymentMethod,
  (products, deliveryOptions, checkout, paymentMethod) =>
    pipe(
      paymentMethod === paymentMethods.creditCard
        ? mapCheckoutDataForCardPayment(products, deliveryOptions)
        : mapCheckoutDataForCOD(products, deliveryOptions),
      paymentMethod === paymentMethods.creditCard
        ? isPayfort
          ? always
          : checkoutCardValidator
        : checkoutCODValidator,
      keys,
      length,
      lt(0)
    )(checkout)
)

const selectIsPayByPayPal = createSelector(
  selectSelectedPaymentMethod,
  equals(paymentMethods.payPal)
)

const formSelector = formValueSelector(form)

const isActiveSelector = createSelector(
  hasCheckoutErrorsSelector,
  selectIsCheckoutPageLocked,
  selectIsCheckoutPageLoading,
  state => !!formSelector(state, 'code'),
  (...args) =>
    pipe(
      any(identity),
      not
    )(args)
)

const PlaceOrderAndPayButton = () => {
  const dispatch = useDispatch()
  const { t } = useTranslation()

  const paymentId = useSelector(selectSelectedPayment)
  const paymentMethod = useSelector(selectSelectedPaymentMethod)
  const selectUserCreditCardProviderByIdSelector = selectUserCreditCardProviderById()
  const provider = useSelector(state =>
    selectUserCreditCardProviderByIdSelector(state, paymentId)
  )
  const isLoading = useSelector(selectIsCheckoutPageLoading)
  const isHidden = useSelector(selectIsPayByPayPal)
  const payByPayFort = useSelector(isPayByPayFort)
  const isActive = useSelector(isActiveSelector)

  const buttonProps = useMemo(
    () => ({
      active: isActive,
      text: isLoading ? t('processing') : t('place_order_and_pay'),
      buttonStyleType: 'primary-dark',
      style: { [s.placeOrderAndPayButton]: true },
      isLoading
    }),
    [isLoading, isActive]
  )

  const onClick = useCallback(
    () => {
      if (payByPayFort && isPayfort) {
        dispatch(CheckoutActions.setIsAddingPaymentCard(true))
      } else if (
        provider === payFortProvider &&
        paymentMethod === paymentMethods.creditCard
      ) {
        dispatch(CheckoutActions.setShowAddCvv(true))
      } else {
        dispatch(CheckoutApiActions.save())
      }
    },
    [payByPayFort, provider, paymentMethod]
  )

  if (isHidden) return null

  return <Button onClick={onClick} {...buttonProps} />
}

export default withStyles(s)(PlaceOrderAndPayButton)
