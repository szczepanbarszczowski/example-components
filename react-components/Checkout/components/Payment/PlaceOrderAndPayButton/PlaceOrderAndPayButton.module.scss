@import 'appBase';

.placeOrderAndPayButton {
  @include my(2);
  text-transform: uppercase;
  font-size: 14px;
  font-weight: 700;
}

.isHidden {
  @include media-breakpoint-down(xs) {
    display: none;
  }
}
