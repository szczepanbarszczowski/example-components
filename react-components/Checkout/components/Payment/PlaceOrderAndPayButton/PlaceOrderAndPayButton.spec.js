import { getReduxComponentWrapper, getDefaultState } from 'appTest'

import { isPayfort } from 'services/configurationCheck'
import { hasCheckoutErrorsSelector } from './PlaceOrderAndPayButton'
import PlaceOrderAndPayButton from './index'

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: () => '' }),
  withTranslation: () => Component => {
    Component.defaultProps = { ...Component.defaultProps, t: () => '' }
    return Component
  }
}))

describe('PlaceOrderAndPayButton', () => {
  it('matches snapshot', () => {
    const wrapper = getReduxComponentWrapper(PlaceOrderAndPayButton, {})
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('errors if no data', () => {
    const result = hasCheckoutErrorsSelector(getDefaultState())
    expect(result).toBe(isPayfort ? false : true)
  })
})
