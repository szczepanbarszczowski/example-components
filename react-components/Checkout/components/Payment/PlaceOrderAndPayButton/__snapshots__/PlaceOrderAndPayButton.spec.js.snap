// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceOrderAndPayButton matches snapshot 1`] = `
<button
  className="buttonWrapper style-primary-dark size-base placeOrderAndPayButton"
  disabled={false}
  onClick={[Function]}
  type="button"
/>
`;
